crates/risc0-boundless/src/host.rs

```diff
@@ -0,0 +1,539 @@
+use tokio::time::{sleep, Duration};
+use tracing::{debug, error, info, instrument, trace, warn};
+use url::Url;
+
+// TODO: Remove blocking wrapper once the boundless client has blocking api
+#[derive(Clone)]
+enum BoundlessRequest {
+    UploadImg {
+        elf: Vec<u8>,
+        notify: UnboundedSender<String>,
+    },
+    UploadInput {
+        input: Vec<u8>,
+        notify: UnboundedSender<String>,
+    },
+    SubmitReq {
+        proving_request: ProvingRequest,
+        notify: UnboundedSender<U256>,
+    },
+    WaitForReqFulfillment {
+        request_id: U256,
+        check_interval: std::time::Duration,
+        timeout: Option<std::time::Duration>,
+        notify: UnboundedSender<Result<(Bytes, Bytes), ClientError>>,
+    },
+    Slash {
+        request_id: U256,
+        notify: UnboundedSender<Result<U256, ClientError>>,
+    },
+}
+
+#[derive(Clone)]
+struct BoundlessClient {
+    queue: tokio::sync::mpsc::UnboundedSender<BoundlessRequest>,
+    _join_handle: Arc<tokio::task::JoinHandle<()>>,
+}
+
+impl BoundlessClient {
+    async fn from_parts(
+        requestor_private_key: PrivateKeySigner,
+        rpc_url: Url,
+        proof_market_address: Address,
+        set_verifier_address: Address,
+    ) -> Self {
+        macro_rules! unwrap_boundless_response {
+            ($response:expr, $queue_loop:lifetime) => (
+                match $response.await {
+                    Ok(r) => r,
+                    Err(e) => {
+                        use ::boundless_market::sdk::client::ClientError;
+                        match e {
+                            // Match against the ClientError enum variants
+                            ClientError::MarketError(e) => {
+                                error!(?e, "Boundless Market Error");
+                                sleep(Duration::from_secs(5)).await;
+                                continue $queue_loop
+                            }
+                            ClientError::StorageProviderError(s) => {
+                                warn!(%s, "Boundless Storage Provider Error");
+                                sleep(Duration::from_secs(10)).await;
+                                continue $queue_loop
+                            }
+                            ClientError::Error(e) => {
+                                error!(?e, "Boundless Error");
+                                sleep(Duration::from_secs(5)).await;
+                                continue $queue_loop
+                            }
+                            _ => {
+                                error!(?e, "Got unrecoverable error from Boundless");
+                                panic!("Boundless API error: {}", e);
+                            }
+                        }
+                    }
+                }
+            );
+        }
+        let (queue, mut rx) = tokio::sync::mpsc::unbounded_channel();
+        let join_handle = tokio::spawn(async move {
```

Comment:
Move this async block into it's own function... 
```suggestion
        let join_handle = tokio::spawn(handle_response(client))
```